app/api/contests/enter.go

```diff
@@ -67,11 +67,23 @@ func (srv *Service) enter(w http.ResponseWriter, r *http.Request) service.APIErr
 			Select("NOW() AS now, items.duration, items.contest_participants_group_id").
 			WithWriteLock().Take(&itemInfo).Error())
 
-		service.MustNotBeError(store.Exec(`
-			INSERT INTO attempts (group_id, item_id, started_at, creator_id, `+"`order`"+`)
-			SELECT ?, ?, ?, ?, IFNULL((SELECT MAX(`+"`order`"+`) FROM attempts WHERE group_id = ? AND item_id = ? FOR UPDATE), 0) + 1`,
-			qualificationState.groupID, qualificationState.itemID,
-			itemInfo.Now, srv.GetUser(r).GroupID, qualificationState.groupID, qualificationState.itemID).Error())
+		var attemptID int64
+		service.MustNotBeError(store.RetryOnDuplicatePrimaryKeyError(func(retryStore *database.DataStore) error {
```

Comment:
```
1> begin;
2> begin;
1> select ifnull(max(id) + 1, 0) from attempts where participant_id=21;
+------------------------+
| ifnull(max(id) + 1, 0) |
+------------------------+
|                      4 |
+------------------------+
2> select ifnull(max(id) + 1, 0) from attempts where participant_id=21;
+------------------------+
| ifnull(max(id) + 1, 0) |
+------------------------+
|                      4 |
+------------------------+
1> insert into attempts (participant_id, id) values (21,4);
Query OK, 1 row affected (0.00 sec)
1> commit;
Query OK, 0 rows affected (0.01 sec)
2> insert into attempts (participant_id, id) values (21,4);
ERROR 1062 (23000): Duplicate entry '21-4' for key 'PRIMARY'
```
So we still cat get the 'Duplicate entry' error

Reply:
But so, wouldn't it wiser (cleaner) to do the operation with a single `INSERT.. SELECT`. This way I suppose it is atomic.

Reply:
Since we use the new id later, we would need to select it separately (because we cannot get the last insert id for the composite primary key). But yes, that is possible.

Reply:
fixed